### DISCUSSION TITLE
Help debugging "Disconnect reason 7" problem

### DISCUSSION THREAD

**Discussion by sude22** [bertmelis/espMqttClient#133] [Q&A] [answered]
Hello Bert,
It's me again. Your MQTT client is really an excellent thing, years of annoyance about MQTT connections not reconnecting are a thing of the past :)
Because it worked so well for a new project, I have written a HOMIE replacement for all my home automation devices. It only contains what I need, but also what I personally was missing in HOMIE.
The basis of the thing is of course your MQTT client.

Maybe you can help me with a problem: Although the reconnect always works reliably, regardless of whether WiFi or the broker fails, I have a problem.
The broker is a current Mosquitto 2.0.15 in a Docker container on the HomeServer. The WiFi is now running quite stable with several AccessPoints from ExtremNetworks under OpenWRT and B.A.T.M.A.N Mesh.

I keep noticing that some devices have to reconnect to the broker after a while. The disconnect reason is 7.

KeepAlive is set to 30, but I send a status message every 10 seconds (more for testing). 
Changing that values did not help.
Do you have any tips on how to debug the problem?

Best regards,
sude22
P.S. happens with ESP8266 devices...and espMqttClientAsync

**Comment by sude22** [comment]
![grafik](https://github.com/bertmelis/espMqttClient/assets/3779162/7fa6d4de-af7a-4edb-87f0-227380ba44f5)
These graphs are not entirely correct, as the broker was restarted more often at the end. However, it gives an impression of the magnitude of the disconnects we are talking about

**Comment by bertmelis** [comment]
Would you be able to get the broker log? Often it logs why it kicks a client.

**Reply by sude22** [reply]
Just turned on the debug logs on Mosquitto...
Well, there comes a lot.. I will try to find some parts that are relevant.

**Comment by sude22** [comment]
OK, first I changed some things...using the "normal" sync client now, removed some 3rd-party OTA stuff.
What I see now is this:
 `2023-12-16T09:19:06: Client esp8266c5b8d8 has exceeded timeout, disconnecting.`
`2023-12-16T09:19:22: New connection from 192.168.1.32:62316 on port 1883.`
`2023-12-16T09:19:22: New client connected from 192.168.1.32:62316 as esp8266c5b8d8 (p2, c1, k20).`
![grafik](https://github.com/bertmelis/espMqttClient/assets/3779162/a90a9958-3bbc-447e-a287-efed77b49743)
One can clearly see that before the connection is interrupted, the heap memory rapidly decreases. The question is whether this is due to my remaining code or the espMqttClient. I will set up a test project with only the client and observe that.

**Comment by Pablo2048** [comment]
@sude22 check the RSSI or wifi disconnect/reconnect, check if you have some guard to handle MQTT disconnect, or set `EMC_ALLOW_NOT_CONNECTED_PUBLISH` flag to 0 (https://www.emelis.net/espMqttClient/#compile-time-configuration) to avoid the memory exhaustion when disconnected.

**Reply by sude22** [reply]
This my publish function, it checks if the client is connected before publishing:

`uint16_t publish(PGM_P topic, uint8_t qos, bool retain, const char *payload)`
`{`
`        if (!mqttClient.connected())`  
`            return INT16_MIN;`  

`        return mqttClient.publish(prefixMqttTopic(topic), qos, retain, payload);`  
`}`

**Reply by Pablo2048** [reply]
Well, then you have to check the wifi connection status also. As I wrote - start debugging with the RSSI value and the wifi connection status.

**Reply by sude22** [reply]
I will try. 
My assumption was that no matter if MQTT or WiFi is not connected, the client returns connected()==false as status

**Comment by proddy** [comment]
not sure if this is related but in one of my projects we had many issues of reconnection issues with MQTT and it turned it to be a problem with espressif32's WiFi connectivity on Mesh networks (specifically FritzBoxes) due to channel roaming. We solved this by using BSSID or fixing IPs in the client code.

**Reply by sude22** [reply]
Thank you, you've given me some good food for thought. 
I have noticed that there is indeed something wrong with the mesh, some ESPs have been wildly "pushed" from one ap to another. Then it's no wonder that the MQTT connection breaks off. I am now converting everything to unique SSIDs without mesh, the ESPs are fixed to one location anyway. 
Let's see what the long-term measurements show. 
Thank you all for your help so far.

**Comment by sude22** [comment]
A small interim result:
The use of unique SSIDs on the APs has significantly improved the situation. Almost all ESPs are now running for several hours with a stable WiFi AND MQTT connection.
Nevertheless, one out of 8 test devices showed a different behavior again. Same software version, same WiFi as the other device.
![grafik](https://github.com/bertmelis/espMqttClient/assets/3779162/61c29edd-fad0-4e34-9896-d6d40b7ecea2)

I have now compiled the espmqttclient with debug output and hope to gain further insights.

Theoretically, I could live with this status, as the MQTT connection is always re-established properly and there is no ESP reset. But now my curiosity has been aroused to clarify the matter.

**Reply by bertmelis** [reply]
Oddly, I'm happy with the situation. But this indeed sparks also my curiosity.

**Comment by sude22** [comment]
Unfortunately, another case has now occurred in which an ESP cannot reconnect to the broker.
Debugging is not switched on for this ESP.

![grafik](https://github.com/bertmelis/espMqttClient/assets/3779162/2ead789c-2403-4c03-9624-c588c8797150)

You can see in the broker that it establishes a new connection, but then the broker times out after 30 seconds.
This then repeats itself again and again.

![grafik](https://github.com/bertmelis/espMqttClient/assets/3779162/88110d23-b58c-4f68-94bd-d6444a7b2e86)


I use the reconnect code from the basic example.
Hopefully the case will also occur once on the device with debugging activated...

**Comment by bertmelis** [comment]
probably related to #106. That user could switch to the sync version. On esp8266 this isn't always the case. ESPAsyncTCP doesn't seem to be the most stable library...

**Comment by sude22** [comment]
Well, i did that [yesterday](https://github.com/bertmelis/espMqttClient/discussions/133#discussioncomment-7870096)..
This is really strange, but i guess without the debug logs there is no way to find out.

**Reply by bertmelis** [reply]
I added state change logging in the latest code to debug stuff like this. Now we don't know in what state of the FSM it hangs.
Attention: It's in the git version, not the released version.

**Reply by sude22** [reply]
OK, then I should try that in the "debug device".. main or branch?

**Reply by bertmelis** [reply]
main

**Reply by sude22** [reply]
BTW, how do you avoid the 
`.pio\libdeps\d1_mini\ESPAsyncTCP\src/DebugPrintMacros.h:23:18: error: 'millis' was not declared in this scope`
error on a fresh build with 
```
build_flags = -D DEBUG_ESP_MQTT_CLIENT=1
              -D CORE_DEBUG_LEVEL=ARDUHAL_LOG_LEVEL_VERBOSE
              -D DEBUG_ESP_PORT=Serial
```

**Reply by sude22** [reply]
I hope that is correct then:
```
16:24:38.798 > Alloc (l:27)
16:24:38.798 > state 0 --> 1
16:24:38.798 > state 1 --> 2
16:24:38.814 > state 2 --> 3
16:24:38.814 > tx 27/27 (10)
16:24:38.824 > rx len 4
16:24:38.824 > Packet type: 0x20
16:24:38.824 > Remaining length: 2
16:24:38.824 > Packet complete
16:24:38.824 > state 3 --> 4
16:24:38.824 > clearing queue (clear session: 1)
16:24:38.824 > Connected to MQTT.
```

**Reply by bertmelis** [reply]
Missing millis is an error from ESPAsyncTCP. Arduino IDE automatically adds `#include <Arduino.h>` to every file but other IDEs might not do that.

Log seems correct. It cycles through the different stages of connecting.

**Reply by sude22** [reply]
Ah, OK. I use VS Code / PlatformIO. 
The debug device is up and running, let's see what happens over night.

**Reply by sude22** [reply]
It is really strange. Nearly all devices of 8 run 24h with only 2 MQTT reconnects over all. Pretty stable, I guess it must have something to do with my WiFi equipment.

But, at least, the debug device lost it's connection and I can provide the broker and ESP-Logs.
Please don't mind, in the ESP debug log the "MqttConnected" millis are always the same as "WifiConnected", c&p mistake.
[disconnect_logs_20231218.zip](https://github.com/bertmelis/espMqttClient/files/13707367/disconnect_logs_20231218.zip)

**Reply by sude22** [reply]
It is in a state that it could not reconnect to the broker again....
[no_reconnect20231218.txt](https://github.com/bertmelis/espMqttClient/files/13707405/no_reconnect20231218.txt)
[EDIT]
Not correct, after a while it connected again. Maybe you can see something in the logs....

**Reply by sude22** [reply, marked as answer]
After all, I really think it is the problem of my special WiFi setup, that sometimes blocks the traffic, even if the WiFi state is connected.

Anyway, thanks for all your time and help with that. I move on with my project, and in case I find out something new about this I let you know.
